Scan support for LongDictionaryStreamReader

oerling opened the pull request:
Moves common functions between direct and dictionary long stream readers to a common superclass.

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java:
```diff
@@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
     {
         int count = 0;
-        for (int i = 0; i < batchSize; i++) {
+        int bitsInFirstByte = Math.min(batchSize, bitsInData);
```
This optimization should go into a separate commit against master.

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java:
```diff
@@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
         return count;
     }
 
+    // vector[i] is set to the bit at offsets[i] - offsetBase for i
```
Method docs should use javadoc style comments, e.g.

```
/**
 * ...
 */
```

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java:
```diff
@@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
+    {
+        if (offsets[0] == offsetBase && offsets[numOffsets - 1] == offsetBase + numOffsets - 1) {
+            getSetBits(numOffsets, vector);
+            if (numBits > offsets[numOffsets - offsetBase - 1] + 1) {
```
How does this calculation work? It seems strange to mix numOffsets and offsetBase in a single calculation. Can't this be simplified to:

```
if (numBits > numOffsets) {
    skip(numBits - numOffsets);
}
```

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java:
```diff
@@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
+        }
+        int position = 0;
+        for (int i = 0; i < numOffsets; i++) {
+            int target = offsets[i] - offsetBase;
```
I would rewrite this piece in terms of previous offset and delta:

```
        int previousOffset = offsetBase;
        for (int i = 0; i < numOffsets; i++) {
            int delta = offsets[i] - previousOffset;
            if (delta > 0) {
                skip(delta);
            }
            vector[i] = nextBit();
            previousOffset = offsets[i] + 1;
        }
```

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java:
```diff
@@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
+            position++;
+        }
+        if (numBits > offsets[numOffsets - 1] - offsetBase) {
+            skip(offsets[numOffsets - 1] - offsetBase - position);
```
Shouldn't this be

```
        if (numBits > offsets[numOffsets - 1] - offsetBase) {
            skip(numBits - (offsets[numOffsets - 1] - offsetBase));
        }
```

oerling commented:
@mbasmanova requested changes on this pull request.

@oerling Overall looks good, but how did you test this?

 

Test with data written by Spark. How do we make more of this? Where do you want to keep it?

 

 

In presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java:

> +    @Override

+    public void erase(int end)

+    {

+        if (values == null) {

+            return;

+        }

+        numValues -= end;

+        if (numValues > 0) {

+            System.arraycopy(values, end, values, 0, numValues);

+            if (valueIsNull != null) {

+                System.arraycopy(valueIsNull, end, valueIsNull, 0, numValues);

+            }

+        }

+    }

+

+    AbstractLongStreamReader()

move constructor before erase

 

In presto-orc/src/main/java/com/facebook/presto/orc/reader/ColumnReader.java:

> @@ -341,4 +341,9 @@ protected void checkEnoughValues(int numFirstRows)

             throw new IllegalArgumentException(format("Reader does not have enough rows: requested %s, available %s", numFirstRows, numValues));

         }

     }

+

+    public static int roundupSize(int size)

This method is used only in LongDictionaryStreamReader, hence, it belongs there.

 

This is mostly needed for repeated, where seen up to 10x garbage multiplication from  1% resizes. So this would be moved here at the next PR. Since the state between this and repeated readers will not go anywhere this may as well be here.

 

 

 

In presto-orc/src/main/java/com/facebook/presto/orc/reader/ColumnReader.java:

> @@ -341,4 +341,9 @@ protected void checkEnoughValues(int numFirstRows)

             throw new IllegalArgumentException(format("Reader does not have enough rows: requested %s, available %s", numFirstRows, numValues));

         }

     }

+

+    public static int roundupSize(int size)

+    {

+        return size + (size / 4);

Where does 1/4 come from? Let's make a constant.

 

OK

In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java:

 

 

OK

 

 

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Tuesday, April 2, 2019 11:27 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Scan support for LongDictionaryStreamReader (#12564)

 

@mbasmanova requested changes on this pull request.

@oerling <https://github.com/oerling>  Overall looks good, but how did you test this?

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271383638> :

> +    @Override
+    public void erase(int end)
+    {
+        if (values == null) {
+            return;
+        }
+        numValues -= end;
+        if (numValues > 0) {
+            System.arraycopy(values, end, values, 0, numValues);
+            if (valueIsNull != null) {
+                System.arraycopy(valueIsNull, end, valueIsNull, 0, numValues);
+            }
+        }
+    }
+
+    AbstractLongStreamReader()

move constructor before erase

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/ColumnReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271384238> :

> @@ -341,4 +341,9 @@ protected void checkEnoughValues(int numFirstRows)
             throw new IllegalArgumentException(format("Reader does not have enough rows: requested %s, available %s", numFirstRows, numValues));
         }
     }
+
+    public static int roundupSize(int size)

This method is used only in LongDictionaryStreamReader, hence, it belongs there.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/ColumnReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271384389> :

> @@ -341,4 +341,9 @@ protected void checkEnoughValues(int numFirstRows)
             throw new IllegalArgumentException(format("Reader does not have enough rows: requested %s, available %s", numFirstRows, numValues));
         }
     }
+
+    public static int roundupSize(int size)
+    {
+        return size + (size / 4);

Where does 1/4 come from? Let's make a constant.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDirectStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271404122> :

> @@ -204,6 +171,7 @@ public void scan()
         QualifyingSet input = hasNulls ? innerQualifyingSet : inputQualifyingSet;
         // Read dataStream if there are non-null values in the QualifyingSet.
         if (input.getPositionCount() > 0) {
+            resultsProcessor.reset();

This change is unnecessary.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271407863> :

> +            numResults = 0;
+            if (filter != null) {
+                rowNumbers = inputQualifyingSet.getPositions();
+                inputNumbers = hasNulls ? innerQualifyingSet.getInputNumbers() : null;
+                rowNumbersOut = outputQualifyingSet.getPositions();
+                inputNumbersOut = outputQualifyingSet.getInputNumbers();
+            }
+            else {
+                rowNumbers = null;
+                inputNumbers = null;
+                rowNumbersOut = null;
+                inputNumbersOut = null;
+            }
+        }
+
+        long adjustValue(int offsetIndex, long value)

private

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java <https://github.com/prestodb/presto/pull/12564#discussion_r271431936> :

> @@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
             throws IOException
     {
         int count = 0;
-        for (int i = 0; i < batchSize; i++) {
+        int bitsInFirstByte = Math.min(batchSize, bitsInData);

This optimization should go into a separate commit against master.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java <https://github.com/prestodb/presto/pull/12564#discussion_r271432534> :

>              vector[i] = nextBit();
             count += vector[i] ? 1 : 0;
         }
         return count;
     }
 
+    // vector[i] is set to the bit at offsets[i] - offsetBase for i

Method docs should use javadoc style comments, e.g.

/**
 * ...
 */
  _____  


In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java <https://github.com/prestodb/presto/pull/12564#discussion_r271433273> :

>              vector[i] = nextBit();
             count += vector[i] ? 1 : 0;
         }
         return count;
     }
 
+    // vector[i] is set to the bit at offsets[i] - offsetBase for i
+    // from 0 to numOffsets. The offset[i] - offsetBase values are
+    // relative to the current position of the stream. 0 would mean
+    // the value returned by a call to next(). The stream is advanced
+    // by numBits. The offset values are increasing and non-repeating.
+    public void getSetBits(int[] offsets, int numOffsets, int offsetBase, int numBits, boolean[] vector)
+            throws IOException
+    {
+        if (offsets[0] == offsetBase && offsets[numOffsets - 1] == offsetBase + numOffsets - 1) {
+            getSetBits(numOffsets, vector);
+            if (numBits > offsets[numOffsets - offsetBase - 1] + 1) {

How does this calculation work? It seems strange to mix numOffsets and offsetBase in a single calculation. Can't this be simplified to:

if (numBits > numOffsets) {
    skip(numBits - numOffsets);
}
  _____  


In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java <https://github.com/prestodb/presto/pull/12564#discussion_r271437599> :

> +    // relative to the current position of the stream. 0 would mean
+    // the value returned by a call to next(). The stream is advanced
+    // by numBits. The offset values are increasing and non-repeating.
+    public void getSetBits(int[] offsets, int numOffsets, int offsetBase, int numBits, boolean[] vector)
+            throws IOException
+    {
+        if (offsets[0] == offsetBase && offsets[numOffsets - 1] == offsetBase + numOffsets - 1) {
+            getSetBits(numOffsets, vector);
+            if (numBits > offsets[numOffsets - offsetBase - 1] + 1) {
+                skip(numBits - offsets[numOffsets - 1] - offsetBase - 1);
+            }
+            return;
+        }
+        int position = 0;
+        for (int i = 0; i < numOffsets; i++) {
+            int target = offsets[i] - offsetBase;

I would rewrite this piece in terms of previous offset and delta:

        int previousOffset = offsetBase;
        for (int i = 0; i < numOffsets; i++) {
            int delta = offsets[i] - previousOffset;
            if (delta > 0) {
                skip(delta);
            }
            vector[i] = nextBit();
            previousOffset = offsets[i] + 1;
        }
  _____  


In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java <https://github.com/prestodb/presto/pull/12564#discussion_r271437768> :

> +                skip(numBits - offsets[numOffsets - 1] - offsetBase - 1);
+            }
+            return;
+        }
+        int position = 0;
+        for (int i = 0; i < numOffsets; i++) {
+            int target = offsets[i] - offsetBase;
+            if (target > position) {
+                skip(target - position);
+                position = target;
+            }
+            vector[i] = nextBit();
+            position++;
+        }
+        if (numBits > offsets[numOffsets - 1] - offsetBase) {
+            skip(offsets[numOffsets - 1] - offsetBase - position);

Shouldn't this be

        if (numBits > offsets[numOffsets - 1] - offsetBase) {
            skip(numBits - (offsets[numOffsets - 1] - offsetBase));
        }

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/12564#pullrequestreview-221758780> , or mute the thread <https://github.com/notifications/unsubscribe-auth/Ap73zypJZqMXKtWPD6qeQoZBFkA6dletks5vc6DugaJpZM4cWP0X> .

mbasmanova commented:
@oerling @tdcmeehan 

> Test with data written by Spark. How do we make more of this? Where do you want to keep it?

How about generating some test data and writing it two ways - with long-dictionary encoding and long-direct encoding - into two resource files (put them under test/resources)? The test then reads two files and ensures that the data is the same.  

See com.facebook.presto.orc.TestFlatMap (thanks to Tim for the pointer)

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java:
```diff
@@ -0,0 +1,110 @@
+        if (values == null) {
+            return;
+        }
+        numValues -= end;
```
numberOfValues or valuesCount ? i.e. the guideline for using full words.

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Override
+    public void compactValues(int[] positions, int base, int numPositions)
```
ditto for numPositions. Maybe length? or positionsCount?

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Override
+    protected void writeNull(int i)
```
Change `i` to `ignored` - I see that convention in the code.

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Override
+    public Block getBlock(int numFirstRows, boolean mayReuse)
```
ditto for `numFirstRows` - how about `initialRowCount`?

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java:
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
+            if (inDictionaryFlags == null || inDictionaryFlags.length < numInput) {
+                inDictionaryFlags = new boolean[roundupSize(numInput)];
+            }
+            int begin = hasNulls ? innerPosInRowGroup : posInRowGroup;
```
`innerPosInRowGroup` -> `innerPositionInRowGroup`
`posInRowGroup` -> `positionInRowGroup`

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java:
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
+        makeInnerQualifyingSet();
+
+        QualifyingSet input = hasNulls ? innerQualifyingSet : inputQualifyingSet;
+        int numInput = input.getPositionCount();
```
`numInput` -> `inputPositionCount` or something similar

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java:
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
+            implements LongInputStream.ResultsConsumer
+    {
+        private int[] offsets;
+        private int[] rowNumbers;
```
Add comments describing what these parameters are used for.

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java:
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
+    {
+        private int[] offsets;
+        private int[] rowNumbers;
+        private int[] inputNumbers;
```
ditto

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java:
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
+        private int[] offsets;
+        private int[] rowNumbers;
+        private int[] inputNumbers;
+        private int[] rowNumbersOut;
```
ditto

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java:
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
+        private int[] rowNumbers;
+        private int[] inputNumbers;
+        private int[] rowNumbersOut;
+        private int[] inputNumbersOut;
```
ditto

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java:
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
+        private int[] inputNumbers;
+        private int[] rowNumbersOut;
+        private int[] inputNumbersOut;
+        private int numResults;
```
`numResults` -> `numberOfResults` or `resultsCount` ?

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java:
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
+                    inputNumbersOut[numResults] = offsetIndex;
+                }
+                else {
+                    int outerIdx = inputNumbers[offsetIndex];
```
`outerIndex`

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java:
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
+                }
+            }
+            if (values != null) {
+                values[numResults + numValues] = adjustValue(offsetIndex, value);
```
`numValues` -> `valuesCount` or `numberOfValues` ?

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java:
```diff
@@ -132,13 +133,66 @@ public int getSetBits(int batchSize, boolean[] vector)
+        }
+        verify(bitsInData == 0);
+        int wholeBytes = (batchSize - bitsInFirstByte) / 8;
+        int numFilled = bitsInFirstByte;
```
`numFilled` to `positionsFilled` or something similar

oerling commented:
Using Num as short for number of is pre-existing practice in the ORC code. Similarly the field names in result consumer are this way in all of them.

If you seriously would consider changing the practice this should be applied throughout. This is best done as a batch rename at the end of everything else. 

 

 

 

 

 

From: Elon Azoulay <notifications@github.com> 
Sent: Tuesday, April 2, 2019 2:31 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Scan support for LongDictionaryStreamReader (#12564)

 

@elonazoulay requested changes on this pull request.

Mostly just comments about variable names otherwise LGTM!

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271500154> :

> +            extends NullWrappingColumnReader
+{
+    protected long[] values;
+
+    AbstractLongStreamReader()
+    {
+        super(OptionalInt.of(SIZE_OF_LONG));
+    }
+
+    @Override
+    public void erase(int end)
+    {
+        if (values == null) {
+            return;
+        }
+        numValues -= end;

numberOfValues or valuesCount ? i.e. the guideline for using full words.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271500453> :

> +    public void erase(int end)
+    {
+        if (values == null) {
+            return;
+        }
+        numValues -= end;
+        if (numValues > 0) {
+            System.arraycopy(values, end, values, 0, numValues);
+            if (valueIsNull != null) {
+                System.arraycopy(valueIsNull, end, valueIsNull, 0, numValues);
+            }
+        }
+    }
+
+    @Override
+    public void compactValues(int[] positions, int base, int numPositions)

ditto for numPositions. Maybe length? or positionsCount?

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271500604> :

> +    {
+        if (outputChannelSet) {
+            StreamReaders.compactArrays(positions, base, numPositions, values, valueIsNull);
+            numValues = base + numPositions;
+        }
+        compactQualifyingSet(positions, numPositions);
+    }
+
+    @Override
+    protected void shiftUp(int from, int to)
+    {
+        values[to] = values[from];
+    }
+
+    @Override
+    protected void writeNull(int i)

Change i to ignored - I see that convention in the code.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271500902> :

> +    }
+
+    @Override
+    protected void shiftUp(int from, int to)
+    {
+        values[to] = values[from];
+    }
+
+    @Override
+    protected void writeNull(int i)
+    {
+        // No action. values[i] is undefined if valueIsNull[i] == true.
+    }
+
+    @Override
+    public Block getBlock(int numFirstRows, boolean mayReuse)

ditto for numFirstRows - how about initialRowCount?

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271501780> :

> +            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+        beginScan(presentStream, null);
+        ensureValuesCapacity();
+        makeInnerQualifyingSet();
+
+        QualifyingSet input = hasNulls ? innerQualifyingSet : inputQualifyingSet;
+        int numInput = input.getPositionCount();
+        if (inDictionaryStream != null) {
+            if (inDictionaryFlags == null || inDictionaryFlags.length < numInput) {
+                inDictionaryFlags = new boolean[roundupSize(numInput)];
+            }
+            int begin = hasNulls ? innerPosInRowGroup : posInRowGroup;

innerPosInRowGroup -> innerPositionInRowGroup
posInRowGroup -> positionInRowGroup

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271501988> :

> @@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
     {
         return INSTANCE_SIZE + sizeOf(nullVector) + sizeOf(dataVector) + sizeOf(inDictionaryVector);
     }
+
+    @Override
+    public void scan()
+            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+        beginScan(presentStream, null);
+        ensureValuesCapacity();
+        makeInnerQualifyingSet();
+
+        QualifyingSet input = hasNulls ? innerQualifyingSet : inputQualifyingSet;
+        int numInput = input.getPositionCount();

numInput -> inputPositionCount or something similar

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271502275> :

> +
+        if (hasNulls) {
+            innerPosInRowGroup = innerQualifyingSet.getEnd();
+        }
+        addNullsAfterScan(filter != null ? outputQualifyingSet : inputQualifyingSet, inputQualifyingSet.getEnd());
+        if (filter != null) {
+            outputQualifyingSet.setEnd(inputQualifyingSet.getEnd());
+        }
+        endScan(presentStream);
+    }
+
+    private final class ResultsProcessor
+            implements LongInputStream.ResultsConsumer
+    {
+        private int[] offsets;
+        private int[] rowNumbers;

Add comments describing what these parameters are used for.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271502294> :

> +        if (hasNulls) {
+            innerPosInRowGroup = innerQualifyingSet.getEnd();
+        }
+        addNullsAfterScan(filter != null ? outputQualifyingSet : inputQualifyingSet, inputQualifyingSet.getEnd());
+        if (filter != null) {
+            outputQualifyingSet.setEnd(inputQualifyingSet.getEnd());
+        }
+        endScan(presentStream);
+    }
+
+    private final class ResultsProcessor
+            implements LongInputStream.ResultsConsumer
+    {
+        private int[] offsets;
+        private int[] rowNumbers;
+        private int[] inputNumbers;

ditto

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271502319> :

> +            innerPosInRowGroup = innerQualifyingSet.getEnd();
+        }
+        addNullsAfterScan(filter != null ? outputQualifyingSet : inputQualifyingSet, inputQualifyingSet.getEnd());
+        if (filter != null) {
+            outputQualifyingSet.setEnd(inputQualifyingSet.getEnd());
+        }
+        endScan(presentStream);
+    }
+
+    private final class ResultsProcessor
+            implements LongInputStream.ResultsConsumer
+    {
+        private int[] offsets;
+        private int[] rowNumbers;
+        private int[] inputNumbers;
+        private int[] rowNumbersOut;

ditto

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271502354> :

> +        }
+        addNullsAfterScan(filter != null ? outputQualifyingSet : inputQualifyingSet, inputQualifyingSet.getEnd());
+        if (filter != null) {
+            outputQualifyingSet.setEnd(inputQualifyingSet.getEnd());
+        }
+        endScan(presentStream);
+    }
+
+    private final class ResultsProcessor
+            implements LongInputStream.ResultsConsumer
+    {
+        private int[] offsets;
+        private int[] rowNumbers;
+        private int[] inputNumbers;
+        private int[] rowNumbersOut;
+        private int[] inputNumbersOut;

ditto

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271502528> :

> +        addNullsAfterScan(filter != null ? outputQualifyingSet : inputQualifyingSet, inputQualifyingSet.getEnd());
+        if (filter != null) {
+            outputQualifyingSet.setEnd(inputQualifyingSet.getEnd());
+        }
+        endScan(presentStream);
+    }
+
+    private final class ResultsProcessor
+            implements LongInputStream.ResultsConsumer
+    {
+        private int[] offsets;
+        private int[] rowNumbers;
+        private int[] inputNumbers;
+        private int[] rowNumbersOut;
+        private int[] inputNumbersOut;
+        private int numResults;

numResults -> numberOfResults or resultsCount ?

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271502815> :

> +                }
+                addResult(offsetIndex + i, value);
+                added++;
+            }
+            return added;
+        }
+
+        private void addResult(int offsetIndex, long value)
+        {
+            if (rowNumbersOut != null) {
+                if (inputNumbers == null) {
+                    rowNumbersOut[numResults] = offsets[offsetIndex];
+                    inputNumbersOut[numResults] = offsetIndex;
+                }
+                else {
+                    int outerIdx = inputNumbers[offsetIndex];

outerIndex

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271503031> :

> +
+        private void addResult(int offsetIndex, long value)
+        {
+            if (rowNumbersOut != null) {
+                if (inputNumbers == null) {
+                    rowNumbersOut[numResults] = offsets[offsetIndex];
+                    inputNumbersOut[numResults] = offsetIndex;
+                }
+                else {
+                    int outerIdx = inputNumbers[offsetIndex];
+                    rowNumbersOut[numResults] = rowNumbers[outerIdx];
+                    inputNumbersOut[numResults] = outerIdx;
+                }
+            }
+            if (values != null) {
+                values[numResults + numValues] = adjustValue(offsetIndex, value);

numValues -> valuesCount or numberOfValues ?

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java <https://github.com/prestodb/presto/pull/12564#discussion_r271503261> :

> @@ -132,13 +133,66 @@ public int getSetBits(int batchSize, boolean[] vector)
             throws IOException
     {
         int count = 0;
-        for (int i = 0; i < batchSize; i++) {
+        int bitsInFirstByte = Math.min(batchSize, bitsInData);
+        for (int i = 0; i < bitsInFirstByte; i++) {
+            vector[i] = nextBit();
+            count += vector[i] ? 1 : 0;
+        }
+        verify(bitsInData == 0);
+        int wholeBytes = (batchSize - bitsInFirstByte) / 8;
+        int numFilled = bitsInFirstByte;

numFilled to positionsFilled or something similar

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/12564#pullrequestreview-221908161> , or mute the thread <https://github.com/notifications/unsubscribe-auth/Ap73z0OKaCOtTScAeC2v2mT9fmhKk69zks5vc8wbgaJpZM4cWP0X> .

oerling commented:
Changed.

The size computation has uses in other places so we do not move it to subclass and then back to superclass.

 

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Tuesday, April 2, 2019 11:27 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Scan support for LongDictionaryStreamReader (#12564)

 

@mbasmanova requested changes on this pull request.

@oerling <https://github.com/oerling>  Overall looks good, but how did you test this?

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271383638> :

> +    @Override
+    public void erase(int end)
+    {
+        if (values == null) {
+            return;
+        }
+        numValues -= end;
+        if (numValues > 0) {
+            System.arraycopy(values, end, values, 0, numValues);
+            if (valueIsNull != null) {
+                System.arraycopy(valueIsNull, end, valueIsNull, 0, numValues);
+            }
+        }
+    }
+
+    AbstractLongStreamReader()

move constructor before erase

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/ColumnReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271384238> :

> @@ -341,4 +341,9 @@ protected void checkEnoughValues(int numFirstRows)
             throw new IllegalArgumentException(format("Reader does not have enough rows: requested %s, available %s", numFirstRows, numValues));
         }
     }
+
+    public static int roundupSize(int size)

This method is used only in LongDictionaryStreamReader, hence, it belongs there.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/ColumnReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271384389> :

> @@ -341,4 +341,9 @@ protected void checkEnoughValues(int numFirstRows)
             throw new IllegalArgumentException(format("Reader does not have enough rows: requested %s, available %s", numFirstRows, numValues));
         }
     }
+
+    public static int roundupSize(int size)
+    {
+        return size + (size / 4);

Where does 1/4 come from? Let's make a constant.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDirectStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271404122> :

> @@ -204,6 +171,7 @@ public void scan()
         QualifyingSet input = hasNulls ? innerQualifyingSet : inputQualifyingSet;
         // Read dataStream if there are non-null values in the QualifyingSet.
         if (input.getPositionCount() > 0) {
+            resultsProcessor.reset();

This change is unnecessary.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r271407863> :

> +            numResults = 0;
+            if (filter != null) {
+                rowNumbers = inputQualifyingSet.getPositions();
+                inputNumbers = hasNulls ? innerQualifyingSet.getInputNumbers() : null;
+                rowNumbersOut = outputQualifyingSet.getPositions();
+                inputNumbersOut = outputQualifyingSet.getInputNumbers();
+            }
+            else {
+                rowNumbers = null;
+                inputNumbers = null;
+                rowNumbersOut = null;
+                inputNumbersOut = null;
+            }
+        }
+
+        long adjustValue(int offsetIndex, long value)

private

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java <https://github.com/prestodb/presto/pull/12564#discussion_r271431936> :

> @@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
             throws IOException
     {
         int count = 0;
-        for (int i = 0; i < batchSize; i++) {
+        int bitsInFirstByte = Math.min(batchSize, bitsInData);

This optimization should go into a separate commit against master.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java <https://github.com/prestodb/presto/pull/12564#discussion_r271432534> :

>              vector[i] = nextBit();
             count += vector[i] ? 1 : 0;
         }
         return count;
     }
 
+    // vector[i] is set to the bit at offsets[i] - offsetBase for i

Method docs should use javadoc style comments, e.g.

/**
 * ...
 */
  _____  


In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java <https://github.com/prestodb/presto/pull/12564#discussion_r271433273> :

>              vector[i] = nextBit();
             count += vector[i] ? 1 : 0;
         }
         return count;
     }
 
+    // vector[i] is set to the bit at offsets[i] - offsetBase for i
+    // from 0 to numOffsets. The offset[i] - offsetBase values are
+    // relative to the current position of the stream. 0 would mean
+    // the value returned by a call to next(). The stream is advanced
+    // by numBits. The offset values are increasing and non-repeating.
+    public void getSetBits(int[] offsets, int numOffsets, int offsetBase, int numBits, boolean[] vector)
+            throws IOException
+    {
+        if (offsets[0] == offsetBase && offsets[numOffsets - 1] == offsetBase + numOffsets - 1) {
+            getSetBits(numOffsets, vector);
+            if (numBits > offsets[numOffsets - offsetBase - 1] + 1) {

How does this calculation work? It seems strange to mix numOffsets and offsetBase in a single calculation. Can't this be simplified to:

if (numBits > numOffsets) {
    skip(numBits - numOffsets);
}
  _____  


In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java <https://github.com/prestodb/presto/pull/12564#discussion_r271437599> :

> +    // relative to the current position of the stream. 0 would mean
+    // the value returned by a call to next(). The stream is advanced
+    // by numBits. The offset values are increasing and non-repeating.
+    public void getSetBits(int[] offsets, int numOffsets, int offsetBase, int numBits, boolean[] vector)
+            throws IOException
+    {
+        if (offsets[0] == offsetBase && offsets[numOffsets - 1] == offsetBase + numOffsets - 1) {
+            getSetBits(numOffsets, vector);
+            if (numBits > offsets[numOffsets - offsetBase - 1] + 1) {
+                skip(numBits - offsets[numOffsets - 1] - offsetBase - 1);
+            }
+            return;
+        }
+        int position = 0;
+        for (int i = 0; i < numOffsets; i++) {
+            int target = offsets[i] - offsetBase;

I would rewrite this piece in terms of previous offset and delta:

        int previousOffset = offsetBase;
        for (int i = 0; i < numOffsets; i++) {
            int delta = offsets[i] - previousOffset;
            if (delta > 0) {
                skip(delta);
            }
            vector[i] = nextBit();
            previousOffset = offsets[i] + 1;
        }
  _____  


In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java <https://github.com/prestodb/presto/pull/12564#discussion_r271437768> :

> +                skip(numBits - offsets[numOffsets - 1] - offsetBase - 1);
+            }
+            return;
+        }
+        int position = 0;
+        for (int i = 0; i < numOffsets; i++) {
+            int target = offsets[i] - offsetBase;
+            if (target > position) {
+                skip(target - position);
+                position = target;
+            }
+            vector[i] = nextBit();
+            position++;
+        }
+        if (numBits > offsets[numOffsets - 1] - offsetBase) {
+            skip(offsets[numOffsets - 1] - offsetBase - position);

Shouldn't this be

        if (numBits > offsets[numOffsets - 1] - offsetBase) {
            skip(numBits - (offsets[numOffsets - 1] - offsetBase));
        }

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/12564#pullrequestreview-221758780> , or mute the thread <https://github.com/notifications/unsubscribe-auth/Ap73zypJZqMXKtWPD6qeQoZBFkA6dletks5vc6DugaJpZM4cWP0X> .

mbasmanova commented:
@oerling The build is failing:

```
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/travis/build/prestodb/presto/presto-orc/src/test/java/com/facebook/presto/orc/TestDwrfDictionaries.java:[78,41] constructor PageSourceOptions in class com.facebook.presto.spi.PageSourceOptions cannot be applied to given types;
  required: int[],int[],boolean,com.facebook.presto.spi.PageSourceOptions.FilterFunction[],int
  found: int[],int[],boolean,com.facebook.presto.spi.PageSourceOptions.FilterFunction[],boolean,int,int
  reason: actual and formal argument lists differ in length
[INFO] 1 error
```

oerling commented:
Fixed

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Monday, April 22, 2019 1:22 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Scan support for LongDictionaryStreamReader (#12564)

 

@oerling <https://github.com/oerling>  The build is failing:

[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/travis/build/prestodb/presto/presto-orc/src/test/java/com/facebook/presto/orc/TestDwrfDictionaries.java:[78,41] constructor PageSourceOptions in class com.facebook.presto.spi.PageSourceOptions cannot be applied to given types;
  required: int[],int[],boolean,com.facebook.presto.spi.PageSourceOptions.FilterFunction[],int
  found: int[],int[],boolean,com.facebook.presto.spi.PageSourceOptions.FilterFunction[],boolean,int,int
  reason: actual and formal argument lists differ in length
[INFO] 1 error

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/12564#issuecomment-485538950> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT2QSLSOJW3SYKL43J3PRYM6BANCNFSM4HCY7ULQ> .

mbasmanova commented on presto-orc/src/test/java/com/facebook/presto/orc/TestDwrfDictionaries.java:
```diff
@@ -41,9 +42,18 @@
-        Block block1 = readFile("test_dictionaries/string-dictionary.orc", VarcharType.VARCHAR);
-        Block block2 = readFile("test_dictionaries/string-row-group-dictionary.dwrf", VarcharType.VARCHAR);
-        compare(VarcharType.VARCHAR, block1, block2);
+        Block block1 = readFile("test_dictionaries/string-dictionary.orc", VARCHAR);
```
- block1 and block2 are confusing as it is not clear which is being tested. I'd rename these to expected and actual.
- I'd also rename compare method to assertEqualBlock
- I'd also inline null1 and null2 variables

```
    @Test
    public void testSlice()
            throws Exception
    {
        Block expected = readFile("test_dictionaries/string-dictionary.orc", VARCHAR);
        Block actual = readFile("test_dictionaries/string-row-group-dictionary.dwrf", VARCHAR);
        assertEqualBlocks(VARCHAR, expected, actual);
    }

    private static void assertEqualBlocks(Type type, Block expected, Block actual)
    {
        assertEquals(expected.getPositionCount(), actual.getPositionCount());
        for (int i = 0; i < expected.getPositionCount(); i++) {
            assertEquals(actual.isNull(i), expected.isNull(i));
            if (!expected.isNull(i)) {
                assertEquals(type.compareTo(expected, i, actual, i), 0);
            }
        }
    }
```

oerling commented:
Updated this and fix-row-group-dict. No special comments.

 

From: Maria Basmanova <notifications@github.com> 
Sent: Tuesday, April 23, 2019 5:03 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Scan support for LongDictionaryStreamReader (#12564)

 

@mbasmanova requested changes on this pull request.

@oerling <https://github.com/oerling>  Looks good % a few comments.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r277912059> :

> +        }
+        Block block = new LongArrayBlock(numFirstRows, valueIsNull == null ? Optional.empty() : Optional.of(valueIsNull), values);
+        values = null;
+        valueIsNull = null;
+        numValues = 0;
+        return block;
+    }
+
+    protected void ensureValuesCapacity()
+    {
+        if (!outputChannelSet) {
+            return;
+        }
+        int capacity = numValues + inputQualifyingSet.getPositionCount();
+        if (values == null) {
+            values = new long[capacity];

Use the new utility method.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r277912161> :

> @@ -172,7 +181,7 @@ public Block readBlock(Type type)
         return builder.build();
     }
 
-    private void openRowGroup()
+    protected void openRowGroup()

@override <https://github.com/override> 

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r277912282> :

>  {
     private static final int INSTANCE_SIZE = ClassLayout.parseClass(LongDictionaryStreamReader.class).instanceSize();
 
+    private static final byte FILTER_NOT_EVALUATED = 0;
+    private static final byte FILTER_PASSED = 1;
+    private static final byte FILTER_FAILED = 2;
+
     private final StreamDescriptor streamDescriptor;
 
     private int readOffset;
     private int nextBatchSize;
 
     private InputStreamSource<BooleanInputStream> presentStreamSource = missingStreamSource(BooleanInputStream.class);

presentStreamSource is already present in ColumnReader; remove

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r277912665> :

> +        int numInput = input.getPositionCount();
+        if (inDictionaryStream != null) {
+            if (inDictionaryFlags == null || inDictionaryFlags.length < numInput) {
+                inDictionaryFlags = newBooleanArrayForReuse(numInput);
+            }
+            int begin = hasNulls ? innerPosInRowGroup : posInRowGroup;
+            int end = input.getEnd();
+            inDictionaryStream.getSetBits(input.getPositions(), numInput, begin, end - begin, inDictionaryFlags);
+        }
+        if (filter != null) {
+            outputQualifyingSet.reset(numInput);
+        }
+
+        if (numInput > 0) {
+            if (filter != null) {
+                outputQualifyingSet.reset(numInput);

*	this statement is redundant; it is a duplicate of L301
*	once it is removed, resultsProcessor.reset(); becomes common first statement in both branches of if-else and therefore can be moved outside:

        if (numInput > 0) {
            resultsProcessor.reset();
            if (filter != null) {
                numInnerResults = dataStream.scan(input.getPositions(), 0, numInput, input.getEnd(), resultsProcessor);
                outputQualifyingSet.setPositionCount(numInnerResults);
            }
            else {
                numInnerResults = dataStream.scan(input.getPositions(), 0, numInput, input.getEnd(), resultsProcessor);
            }
        }
  _____  


In presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java <https://github.com/prestodb/presto/pull/12564#discussion_r277913753> :

> +
+        @Override
+        public boolean consume(int offsetIndex, long value)
+        {
+            if (filter != null && !test(offsetIndex, value)) {
+                return false;
+            }
+
+            addResult(offsetIndex, value);
+            return true;
+        }
+
+        @Override
+        public int consumeRepeated(int offsetIndex, long value, int count)
+        {
+            boolean filterDone = filter == null;

same comments as for similar logic in SliceDictionaryStreamReader

        @Override
        public int consumeRepeated(int offsetIndex, long value, int count)
        {
            if (inDictionaryStream == null && (deterministicFilter || filter == null)) {
                // Common case: One filter and one translation
                // from value to dictionary id is enough if there
                // are no row group dictionaries and the filter is
                // null or deterministic.
                if (filter != null && !test(offsetIndex, value)) {
                    return 0;
                }
                for (int i = 0; i < count; i++) {
                    addResult(offsetIndex + i, value);
                }
                return count;
            }
            int added = 0;
            for (int i = 0; i < count; i++) {
                if (filter != null && !test(offsetIndex + i, value)) {
                    continue;
                }
                addResult(offsetIndex + i, value);
                added++;
            }
            return added;
        }
  _____  


In presto-orc/src/test/java/com/facebook/presto/orc/TestDwrfDictionaries.java <https://github.com/prestodb/presto/pull/12564#discussion_r277914735> :

> @@ -41,9 +42,18 @@
     public void testSlice()
             throws Exception
     {
-        Block block1 = readFile("test_dictionaries/string-dictionary.orc", VarcharType.VARCHAR);
-        Block block2 = readFile("test_dictionaries/string-row-group-dictionary.dwrf", VarcharType.VARCHAR);
-        compare(VarcharType.VARCHAR, block1, block2);
+        Block block1 = readFile("test_dictionaries/string-dictionary.orc", VARCHAR);

*	block1 and block2 are confusing as it is not clear which is being tested. I'd rename these to expected and actual.
*	I'd also rename compare method to assertEqualBlock
*	I'd also inline null1 and null2 variables

    @Test
    public void testSlice()
            throws Exception
    {
        Block expected = readFile("test_dictionaries/string-dictionary.orc", VARCHAR);
        Block actual = readFile("test_dictionaries/string-row-group-dictionary.dwrf", VARCHAR);
        assertEqualBlocks(VARCHAR, expected, actual);
    }
 
    private static void assertEqualBlocks(Type type, Block expected, Block actual)
    {
        assertEquals(expected.getPositionCount(), actual.getPositionCount());
        for (int i = 0; i < expected.getPositionCount(); i++) {
            assertEquals(actual.isNull(i), expected.isNull(i));
            if (!expected.isNull(i)) {
                assertEquals(type.compareTo(expected, i, actual, i), 0);
            }
        }
    }

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/12564#pullrequestreview-229846378> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT2DHYIAHYOMQJ3VVM3PR6PS7ANCNFSM4HCY7ULQ> .

mbasmanova commented:
@oerling This is looking great. Please, rebase on top of aria-scan-research and update the PR.

oerling commented:
Updated

 

From: Maria Basmanova <notifications@github.com> 
Sent: Wednesday, April 24, 2019 4:36 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Scan support for LongDictionaryStreamReader (#12564)

 

@oerling <https://github.com/oerling>  This is looking great. Please, rebase on top of aria-scan-research and update the PR.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/12564#issuecomment-486465417> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT2G6KEU3C25HBSXNRDPSDVG3ANCNFSM4HCY7ULQ> .

oerling commented:
Thanks, I put the constant-columns diff on top of this.

 

 

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Thursday, April 25, 2019 9:21 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Scan support for LongDictionaryStreamReader (#12564)

 

Merged #12564 <https://github.com/prestodb/presto/pull/12564>  into aria-scan-research.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/12564#event-2300781269> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPTZDCA4QAA74DR7A3ELPSHK6HANCNFSM4HCY7ULQ> .

mbasmanova commented:
@oerling 

> Thanks, I put the constant-columns diff on top of this.

Great. I'm going to review that next.
